k2/csrc/fsa_utils.cu

```diff
@@ -1616,27 +1640,24 @@ Array1<FloatType> BackpropGetForwardScores(
           });
     }
   } else {
-    int32_t num_arcs = fsas.NumElements(),
-          num_states = fsas.TotSize(1);
-
     K2_CHECK_NE(entering_arcs, nullptr);
     K2_CHECK_EQ(entering_arcs->Dim(), num_states);
+    K2_CHECK(entering_arcs->Context()->IsCompatible(*c));
     const int32_t *entering_arcs_data = entering_arcs->Data();
 
     for (int32_t b = num_batches - 1; b >= 0; --b) {
       int32_t arc_begin;
       Ragged<int32_t> leaving_arc_batch =
           arc_batches_splitter.GetElement(b, &arc_begin);
       int32_t *leaving_arc_batch_data = leaving_arc_batch.values.Data();
-      int32_t num_arcs = leaving_arc_batch.NumElements();
-
       Ragged<FloatType> leaving_arc_deriv(leaving_arc_batch.shape);
       FloatType *leaving_arc_deriv_data = leaving_arc_deriv.values.Data();
 
       K2_EVAL(
-          c, num_arcs, lambda_set_arc_deriv_etc, (int32_t arc_idx)->void {
+          c, leaving_arc_batch.NumElements(), lambda_set_arc_deriv_etc,
+          (int32_t arc_idx)->void {
             int32_t fsas_arc_idx012 = leaving_arc_batch_data[arc_idx];
-            Arc arc = arcs[fsas_arc_idx012];
+            const Arc &arc = arcs[fsas_arc_idx012];
```

Comment:
Why to use a reference here? Is `Arc arc` using local memory?

Reply:
i think we don't need to copy to local memory here